blog/content/post/2020-08-05-PCIe-features.md

```diff
@@ -0,0 +1,213 @@
+that is known to be possible to be implemented in every Function, but most of them
+are optional. Every capability has registers in a Device mapped to configuration
+space. Those registers are a kind of interface to the capability. Some are read
+only and contain information needed by the software, others are read/write and
+can be used to pass information to the Function.
+
+## List of capabilities structures:
+(This list can be incomplete)
+
+- **PCI Express Capability register block**
+
+This is one of the most important capability structure, it must be present in
+all PCIe Functions. It is a collection of various information about:
+
+1. **Device** e.g.:  
+
+    - Maximum payload of Transaction Layer Packet size that the Function can support,
+    - Maximum accepted change latency between power states.  
+
+2. **Link** e.g.:  
+
+    - Link max speed \[GT/s\]
+    - Link max width (number of lines)
+    - Link change latency between power states
+    - Port number for given Link  
+
+3. **Port** e.g.:
+  
+    - Physical slot number which is a chassis unique identifier for a slot.  
+
+4. **Hot-Plug**
+
+    Registers responsible for this capability are located in the Capability register
+    block. This capability allows change of Device in the PCIe slot at runtime.
+    By writing to registers corresponding to this capability OS can let the Root or
+    Switch Ports know to power Off or power On. After power is turned Off an user
+    can safely remove the Device, after inserting new one power should be turned On
+    to its Port.
+
+5. **Baseline Error Reporting**
+
+    Bits corresponding to this capability are located partially in the configuration
+    header, partially in the Capability register block. All of them must be present in
+    all PCIe Functions. Some bits of this capability are used to set error reporting,
+    others store status of errors.
+
+- **Power Management**
+
+OS can manage the power environment of a Function directly by accessing registers
+corresponding to this capability. OS can set Device state to one of 4 states. 2
+of them (D0, D3) are mandatory and other 2 (D2, D3) can be optionally implemented.
+D0 state is a state where the Function is fully operational and uses full power,
+every next state (D1,D2,D3) uses lower power, but also takes more time to recover
+to D0.
+
+- **Message Signaled Interrupts**
+
+PCI has pins to let the central interrupt controller know that it needs service.
+This improves efficiency of the CPU which does not need to check every Function
+periodically, but it also significantly increases the number of needed lines. In
+PCIe instead of additional lines special messages are used to signal interrupt.
+Message Signaled Interrupts (MSI) allows the Function to write a small amount of
+interrupt-describing data to a special memory-mapped I/O address, and the interrupt
+controller then delivers the corresponding interrupt to a processor. Data and
```

Comment:
, then deliver

Reply:
Why? It is the third person of the present simple tense, so the "s" suffix should be added to the verb. Isn't it? 

But I agree that this sentence is too long. I'm changing it to:
"Message Signaled Interrupts (MSI) allow the Function to write a small amount of
interrupt-describing data to a special memory-mapped I/O address. The interrupt
controller then delivers the corresponding interrupt to a processor."